developer-docs-site/docs/move/book/functions.md

```diff
@@ -521,44 +521,61 @@ fun foo() { return () }
+Some simple examples are: `|x| x + 1`, `|x, y| x + y`, `|| 1`, `|| { 1 }`.
+A lambda's body can refer to variables available in the scope where the lambda is defined: this is also known as capturing.
+Such variables can be read or written (if mutable) by the lambda expression; since a lambda expression currently can only occur as a direct argument to a call to an inline function, there are no complicated scoping issues as in some other languages.
+The type of a function parameter is written as `|<list of parameter types>| <return type>`.
+For example, when the function parameter type is `|u64, u64| bool`, any lambda expression that takes two `u64` parameters and returns a `bool` value can be provided as the argument.
 
-Consider the following function (from the `vector` module):
+Below is an example that showcases many of these concepts in action (this example is taken from the `std::vector` module):
 
 ```move=
-/// Fold the function over the elements. For example, `fold(vector[1,2,3], 0, f)` will execute
-/// `f(f(f(0, 1), 2), 3)`
+/// Fold the function over the elements. 
+/// E.g, `fold(vector[1,2,3], 0, f)` is the same as `f(f(f(0, 1), 2), 3)`.
 public inline fun fold<Accumulator, Element>(
     v: vector<Element>,
     init: Accumulator,
     f: |Accumulator,Element|Accumulator
 ): Accumulator {
   let accu = init;
-  foreach(v, |elem| accu = g(accu, elem));
+  // Note: `for_each` is an inline function, but is not shown here.
+  for_each(v, |elem| accu = f(accu, elem));
   accu
 }
 ```
 
-Here, `foreach` is itself an inline function.
+The type signature of the elided public inline function `for_each` is `fun for_each<Element>(v: vector<Element>, f: |Element|)`.
+Its second parameter `f` is a function parameter which accepts any lambda expression that consumes an `Element` and returns nothing.
+In the code example, we use the lambda expression `|elem| accu = f(accu, elem)` as an argument to this function parameter.
+Note that this lambda expression captures the variable `accu` from the outer scope.
+
+### Current restrictions
+
+There are plans to loosen some of these restrictions in the future, but for now,
 
-In general, only lambda expressions can be passed to function parameters.
-Similar as the inline function itself, those lambdas are expanded at caller
-side. Notice that a lambda can access variables in the context, as in the
-example the variable `accu`.
+- Only inline functions can have function parameters.
+- Only explicit lambda expressions can be passed as an argument to an inline function's function parameters.
+- Inline functions and lambda expressions cannot have `return` expressions.
+- Inline functions or lambda expressions cannot return lambda expressions.
+- Cyclic recursion involving only inline functions is not allowed.
+- Parameters in lambda expressions should not be type annotated (e.g., `|x: u64| x + 1` is not allowed): their types are inferred.
```

Comment:
Are we sure that types in lambda parameters aren't working? I definitely extended the parser to support this for v2, but forgot how v1 deals with this. Perhaps try it out first?

Reply:
Yes. I tried the following code on the main branch:

```
module 0xc0ffee::dummy1 {

    inline fun apply(f: |u64|u64, x: u64): u64 {
        f(x)
    }

    fun applier(): u64 {
        apply(|x: u64| x + 1, 1)
    }
}

```

and got the error: 
```
  │         apply(|x: u64| x + 1, 1)
  │               - ^ Expected '|'
  │               │  
  │               To match this '|'
```

Note that the same parse error shows up for compiler v2 as well.